algebird-core/src/main/scala/com/twitter/algebird/MapAlgebra.scala

```diff
@@ -222,4 +224,31 @@ object MapAlgebra {
 
   def dot[K, V](left: Map[K, V], right: Map[K, V])(implicit mring: Ring[Map[K, V]], mon: Monoid[V]): V =
     Monoid.sum(mring.times(left, right).values)
+
+  def cube[K, V](it: TraversableOnce[(K, V)])(implicit c: Cuber[K]): Map[c.K, TraversableOnce[V]] = {
+    val sg = new Semigroup[Iterable[V]] {
+      def plus(x: Iterable[V], y: Iterable[V]) = x ++ y
+    }
+    cubeSum(it.map { case (k, v) => (k, Iterable(v)) })(c, sg)
+  }
+
+  def cubeSum[K, V](it: TraversableOnce[(K, V)])(implicit c: Cuber[K], sg: Semigroup[V]): Map[c.K, V] =
+    sumByKey(it.toIterator.flatMap { case (k, v) => c(k).map((_, v)) })
+
+  def cubeAggregate[T, K, U, V](it: TraversableOnce[T], agg: Aggregator[T, U, V])(fn: T => K)(implicit c: Cuber[K]): Map[c.K, V] =
+    sumByKey(it.toIterator.flatMap { t => c(fn(t)).map((_, agg.prepare(t))) })(agg.semigroup).mapValues(agg.present)
+
+  def rollup[K, V](it: TraversableOnce[(K, V)])(implicit r: Roller[K]): Map[r.K, TraversableOnce[V]] = {
+    val sg = new Semigroup[Iterable[V]] {
+      def plus(x: Iterable[V], y: Iterable[V]) = x ++ y
+    }
+    rollupSum(it.map { case (k, v) => (k, Iterable(v)) })(r, sg)
+  }
+
+  def rollupSum[K, V](it: TraversableOnce[(K, V)])(implicit r: Roller[K], sg: Semigroup[V]): Map[r.K, V] =
+    sumByKey(it.toIterator.flatMap { case (k, v) => r(k).map((_, v)) })
+
+  def rollupAggregate[T, K, U, V](it: TraversableOnce[T], agg: Aggregator[T, U, V])(fn: T => K)(implicit r: Roller[K]): Map[r.K, V] =
+    sumByKey(it.toIterator.flatMap { t => r(fn(t)).map((_, agg.prepare(t))) })(agg.semigroup).mapValues(agg.present)
```

Comment:
mapValues is lazy in scala, and we tend not to use it (because it keeps things in scope we don't expect, and we recompute the values over and over).